commerce/bookstore/schema.tql

```diff
@@ -0,0 +1,134 @@
+#
+# Copyright (C) 2022 Vaticle
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+define
+
+### Attributes ###
+
+    name sub attribute, value string;
+    date sub attribute, value datetime;
+    description sub attribute, value string;
+    id sub attribute, value string;
+
+    price sub attribute, value long;
+    stock sub attribute, value long;
+
+    ISBN sub attribute, value string;
+    book-author sub attribute, value string;
+    publisher sub attribute, value string;
+
+    rating sub attribute, value long;
+
+    age sub attribute, value long;
+
+    foreign-user-id sub attribute, value string;
+    status sub attribute, value string;
+    delivery-address sub attribute, value string;
+    payment-details sub attribute, value string;
+
+    username sub attribute, value string;
+    password sub attribute, value string;
+    foreign-id sub attribute, value string;
+
+    genre sub attribute, value string,
+        plays taging:tag,
+        plays tag-hierarchy:sup-tag,
+        plays tag-hierarchy:sub-tag;
+
+### Entities ###
+    product sub entity, abstract,
+        owns id @key,
+        owns name,
+        owns description,
+        owns price,
+        owns stock,
+        plays ordering:item,
+        plays reviewing:product;
+
+    book sub product,
+        owns ISBN,
+        owns book-author,
+        owns publisher,
+        plays taging:book;
+
+    person sub entity,
+        owns name,
+        plays reviewing:author;
+
+    user sub person,
+        owns id @key,
+        owns foreign-id,
+        owns username,
+        owns password,
+        owns age,
+        plays ordering:author;
+
+    order sub entity,
+        owns id @key,
+        owns foreign-user-id,
+        owns date,
+        owns status,
+        owns delivery-address,
+        owns payment-details,
+        plays ordering:order;
+
+### Relations ###
+    reviewing sub relation,
+        owns rating,
+        relates author,
+        relates product;
+
+    taging sub relation,
```

Comment:
When it comes to assigning tags to books, there are a couple of questions here to be answered.

1) Is the name `tag` clear and unambiguous? Is using `tag` as a role name a good idea, given how many meanings `tag` has, and given that it could easily be an attribute - or even an entity?
2) Do we need a relation to tag a book - given that the tags themselves (genres) are just string attributes?

I would therefore propose the following remodel.

- Make `book owns genre` directly. This eliminates the need for the `tagging` relation. (Note that `owns` is a one-to-many ownership!) This will also affect the implementation of the `super-tag` rule. It will still work.
- Consider renaming `genre` to `genre-tag`. Sure, intuitively, a genre is a well-established concept, but it does feel like more of an entity. A `genre-tag` sounds more like a piece of string data. Plus, it lets us keep the word "tag" for clarity to readers. And in our program, we never use a `genre` for any reason other than to represent a `genre-tag`.
- Consider renaming `tag-hierarchy` to `genre-tag-hierarchy`. This makes it crystal clear what kinds of tags we store in there. Sure, further down the line we might have different kinds of tags, but for now, they are all genres.

Reply:
Books now own genre attribute - https://github.com/vaticle/typedb-examples/pull/143/commits/a615078c50502310693844ee1e66f98501651f93

Reply:
Renamed genre to genre-tag - https://github.com/vaticle/typedb-examples/pull/143/commits/404caa158e90aa0764b4034d18fd4e34b9c4bbbd

Reply:
Since we are taking into account common logic - tag-hierarchy relation can be applied not only for genre-tag but for any tag. So for the sake of logic and as well for being a shorter name, I will leave the tag-hierarchy relation named as it is now.